### PR TITLE
Print the two types in the span label for transmute errors.

### LINKED ISSUE #37157: transmuting between size can vary
rustc 1.14.0-nightly (a3bc191b5 2016-10-10)

``` rust
pub trait TypeConstructor<'a> {
    type T;
}

unsafe fn transmute_lifetime<'a, 'b, C>(x: <C as TypeConstructor<'a>>::T)
                                        -> <C as TypeConstructor<'b>>::T
where for<'z> C: TypeConstructor<'z> {
    ::std::mem::transmute(x)
}
```

``` rust
error[E0512]: transmute called with differently sized types: <C as TypeConstructor<'a>>::T (size can vary because of <C as TypeConstructor>::T) to <C as TypeConstructor<'b>>::T (size can vary because of <C as TypeConstructor>::T)
  --> src/lib.rs:47:5
   |
47 |     ::std::mem::transmute(x)
   |     ^^^^^^^^^^^^^^^^^^^^^ transmuting between size can vary because of <C as TypeConstructor>::T and size can vary because of <C as TypeConstructor>::T
```

The error message near `^^^^^^` doesn’t seem grammatically correct.

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Fixes #37157. I'm not entirely happy with the changes here but overall it's better in my opinion; we certainly avoid the odd language in that issue, which changes to:

```
error[E0512]: transmute called with differently sized types: <C as TypeConstructor<'a>>::T (size can vary because of <C as TypeConstructor>::T) to <C as TypeConstructor<'b>>::T (size can vary because of <C as TypeConstructor>::T)
 --> test.rs:8:5
  |
8 |     ::std::mem::transmute(x)
  |     ^^^^^^^^^^^^^^^^^^^^^ transmuting between <C as TypeConstructor<'a>>::T and <C as TypeConstructor<'b>>::T

error: aborting due to previous error(s)
```

### CODE REVIEW

File: src/librustc/middle/intrinsicck.rs [hunk 1]
```diff
@@ -126,9 +126,7 @@ impl<'a, 'gcx, 'tcx> ExprVisitor<'a, 'gcx, 'tcx> {
                   from, skeleton_string(from, sk_from),
                   to, skeleton_string(to, sk_to))
             .span_label(span,
-                format!("transmuting between {} and {}",
-                    skeleton_string(from, sk_from),
-                    skeleton_string(to, sk_to)))
+                format!("transmuting between {} and {}", from, to))
```
Comment (nikomatsakis) [inline src/librustc/middle/intrinsicck.rs]: at minimum, this should be `{}` and `{}` -- but it definitely feels like we can do better. If nothing else, these error messages are really long, and I think it'd be better to break it up in parts. Let

**Comment by nikomatsakis** [comment]
Maybe something like this?

```
error[E0512]: transmute called with two types that may have different sizes
 --> test.rs:8:5
  |
8 |     ::std::mem::transmute(x)
  |     ^^^^^^^^^^^^^^^^^^^^^ source and target type of transmute may have different sizes
  |
  | note: source type is `<C as TypeConstructor<'a>>::T` (size can vary because of `<C as TypeConstructor>::T`)
  | note: target type is `<C as TypeConstructor<'b>>::T` (size can vary because of `<C as TypeConstructor>::T`)
```

**Comment by nikomatsakis** [comment]
cc @estebank

**Comment by estebank** [comment]
For the general case, I feel something along the lines of this would be ok:

```
error[E0512]: transmute called with two types that may have different sizes
 --> test.rs:8:5
  |
8 |     ::std::mem::transmute(x)
  |     ^^^^^^^^^^^^^^^^^^^^^ source and target type of transmute may have different sizes
  |
  | note: source type is `<C as TypeConstructor<'a>>::T`
          it's size can vary because of `<C as TypeConstructor>::T`
  | note: target type is `<C as TypeConstructor<'b>>::T`
          it's size can vary because of `<C as TypeConstructor>::T`
```

but I'd like to have a specific message for this case along the lines of

```
error[E0512]: transmute called with two types that may have different sizes
 --> test.rs:8:5
  |
8 |     ::std::mem::transmute(x)
  |     ^^^^^^^^^^^^^^^^^^^^^ source and target type of transmute may have different sizes
  |
  | note: source type is `<C as TypeConstructor<'a>>::T` and target type is `<C as TypeConstructor<'b>>::T`
          their size can vary because of `<C as TypeConstructor>::T`
```

For example, in 

```rust
fn takes_u8(_: u8) {}
  
fn main() {
      unsafe { takes_u8(::std::mem::transmute(0u16)); }
}
```

I feel the current output is relatively ok:

```
error[E0512]: transmute called with differently sized types: u16 (16 bits) to u8 (8 bits)
 --> <anon>:4:25
  |
4 |       unsafe { takes_u8(::std::mem::transmute(0u16)); }
  |                         ^^^^^^^^^^^^^^^^^^^^^ transmuting between 16 bits and 8 bits
```

and would look ridiculous under the above proposal:

```
error[E0512]: transmute called with two types that may have different sizes
 --> <anon>:4:25
  |
4 |       unsafe { takes_u8(::std::mem::transmute(0u16)); }
  |                         ^^^^^^^^^^^^^^^^^^^^^ source and target type of transmute may have different sizes
  | note: source type is `u16`
          16 bits
  | note: target type is `u8`
          8 bits
```

---

Because of all of this, I would check for the case where there's an unknown layout in the source and target. If they are, present my second proposed output, otherwise, @nikomatsakis proposed output. If both source and target are not of unknown layout, either have a single-line note or (my preference) keep the current label.

Part of the problem here is that this is two different errors conceptually, the compiler _knows_ in the last example that the types have different sizes, while the other case they _might_ be of different sizes, so they warrant slightly different output.

**Comment by Mark-Simulacrum** [comment]
Hm, okay. I can't say I entirely follow the discussion quite right now, but I'll look into implementing things in this way soon.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum great let me know if you need anything

**Comment by estebank** [comment]
@Mark-Simulacrum It'd also be great if you also moved these tests to `ui`.

**Comment by nikomatsakis** [comment]
> would look ridiculous under the above proposal

fwiw, I don't think that final version looks particularly ridiculous. In fact, I prefer it. I think in particular I'd like to migrate towards a version where the "main error message" is fairly generic, and the details are exposed elsewhere. I think a lot of times we try to pack too much info into the header. But we really need to try and schedule some time to draw up some overall guidelines in this respect.

File: src/librustc/middle/intrinsicck.rs [hunk 2]
```diff
@@ -92,7 +92,7 @@ impl<'a, 'gcx, 'tcx> ExprVisitor<'a, 'gcx, 'tcx> {
                     struct_span_err!(self.infcx.tcx.sess, span, E0591,
                                      "`{}` is zero-sized and can't be transmuted to `{}`",
                                      from, to)
-                        .span_note(span, "cast with `as` to a pointer instead")
```
Comment (Mark-Simulacrum) [inline src/librustc/middle/intrinsicck.rs]: The span here was entirely not helpful as far as I can tell; but I'm open to re-adding it if there's disagreement.

File: src/librustc/middle/intrinsicck.rs [hunk 3]
```diff
@@ -111,7 +111,7 @@ impl<'a, 'gcx, 'tcx> ExprVisitor<'a, 'gcx, 'tcx> {
                 }
                 Err(LayoutError::Unknown(bad)) => {
                     if bad == ty {
-                        format!("size can vary")
```
Comment (Mark-Simulacrum) [inline src/librustc/middle/intrinsicck.rs]: I'm not sure if there's much point to this branch myself; I can't come up with an example that would follow it.

File: src/test/ui/transmute/transmute-from-fn-item-types-error.stderr [hunk 4]
```diff
@@ -0,0 +1,75 @@
+error[E0591]: `unsafe fn() {bar}` is zero-sized and can't be transmuted to `isize`
+  --> $DIR/transmute-from-fn-item-types-error.rs:14:13
+   |
+14 |     let i = mem::transmute(bar);
+   |             ^^^^^^^^^^^^^^
+   |
+   = help: cast with `as` to a pointer instead
+
+error[E0591]: `unsafe fn() -> (isize, *const (), std::option::Option<fn()>) {foo}` is zero-sized and can't be transmuted to `*const ()`
```
Comment (Mark-Simulacrum) [inline src/test/ui/transmute/transmute-from-fn-item-types-error.stderr]: Should we change this error to also follow the same pattern of a simple header and notes to explain it?

**Comment by Mark-Simulacrum** [comment]
Updated; left a few comments, let me know what you think.

File: src/test/ui/transmute/transmute-from-fn-item-types-error.stderr (hunk 4 above)
Comment (estebank) [inline src/test/ui/transmute/transmute-from-fn-item-types-error.stderr]: I think that would be best. The error text should be something like `can't transmute zero-sized type`.

**Review by estebank — COMMENTED** [review]
I'm ok with the output.

**Comment by Mark-Simulacrum** [comment]
Updated the zero-sized case as well.

File: src/librustc/middle/intrinsicck.rs [hunk 5]
```diff
@@ -121,14 +122,9 @@ impl<'a, 'gcx, 'tcx> ExprVisitor<'a, 'gcx, 'tcx> {
...
         struct_span_err!(self.infcx.tcx.sess, span, E0512,
-                  "transmute called with differently sized types: \
-                   {} ({}) to {} ({})",
-                  from, skeleton_string(from, sk_from),
-                  to, skeleton_string(to, sk_to))
-            .span_label(span,
-                format!("transmuting between {} and {}",
-                    skeleton_string(from, sk_from),
-                    skeleton_string(to, sk_to)))
+            "transmute called with differently sized types")
```
Comment (nikomatsakis) [inline src/librustc/middle/intrinsicck.rs]: Nit: "called with types of different sizes" sounds more grammatical to me

**Comment by nikomatsakis** [comment]
This looks good to me modulo the nit about the error message phrasing.

**Comment by Mark-Simulacrum** [comment]
@bors r=nikomatsakis

Fixed the nit.

**Comment by Mark-Simulacrum** [comment]
@bors r-

I need to fix the UI tests to not be dependent on 32/64-bit (they use usize/isize).

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum in this case, it may suffice to restrict the platforms that they run on, but this is an interesting problem with `ui` tests that has cropped up a few times now. It seems like in the limit we need some way to make more general reference files (i.e., that contain some wildcards), or perhaps to just have per-platform reference files.

**Comment by Mark-Simulacrum** [comment]
I rebased and changed the `isize` and `usize` to `i32` and `u32`; however, I believe that won't help with the function pointers. I'd be happy to ignore some platforms, but I'm not quite sure which... could you help me out?

**Comment by estebank** [comment]
@nikomatsakis filed #42434 for UI wildcard support.

**Comment by Mark-Simulacrum** [comment]
@bors r-

rollup failed, but need to investigate
https://travis-ci.org/rust-lang/rust/builds/239684191

**Comment by Mark-Simulacrum** [comment]
I'm seeing diffs like below on armhf-gnu and dist-i586-gnu-i686-musl, which is odd -- I sort of don't expect types to be sometimes zero sized? @nikomatsakis Should I just ignore these platforms as well, and if so, how -- what `ignore-XXX` do I use? x86 seems to not work, though I would expect it to (or maybe I'm wrong and it does, and I need to add it to the two other transmute tests).

```
[01:20:46] -error[E0512]: transmute called with types of different sizes
[01:20:46] +error[E0591]: can't transmute zero-sized type
[01:20:46]    --> $DIR/transmute-from-fn-item-types-error.rs:16:13
[01:20:46]     |
[01:20:46]  16 |     let i = mem::transmute(bar);
[01:20:46]     |             ^^^^^^^^^^^^^^
[01:20:46]     |
[01:20:46] -   = note: source type: unsafe fn() {bar} (0 bits)
[01:20:46] -   = note: target type: i32 (32 bits)
[01:20:46] +   = note: source type: unsafe fn() {bar}
[01:20:46] +   = note: target type: i32
[01:20:46] +   = help: cast with `as` to a pointer instead
```

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum that is a bit odd, yeah. But it doesn't look like the types are only sometimes zero-sized -- after all, it prints `0 bits`. It seems like whatever check we are using to switch *error messages* is perhaps unreliable?

**Comment by arielb1** [comment]
@Mark-Simulacrum
```
 +                if size_to == Pointer.size(self.tcx) { 
```

i32 is pointer-sized on 32-bit archs, and non-pointer-sized otherwise.

**Comment by Mark-Simulacrum** [comment]
That explains it, I think, though I'm not sure I know how to change the tests to still test this case but not be architecture dependent... I'll need to think on this.

**Comment by arielb1** [comment]
@Mark-Simulacrum 

I think we might want to have tests only run on 32-bit or 64-bit architectures, and have separate paths for that.

**Comment by nikomatsakis** [comment]
certainly the short-term hack is just to have two tests, one for 32-bit and one for 64-bit? But I thought that the idea of being able to add custom "filters" into the `ui` tests was a nice long-term solution. Not sure where that issue ran off to.

**Comment by Mark-Simulacrum** [comment]
Yeah, I hope to get around to adding the two tests later this week, I just need to figure out the specific syntax and what architectures I need to ignore.

**Comment by Mark-Simulacrum** [comment]
I just ignored the tests on 32-bit targets. I personally don't know if it's worth testing on those, the tests are just for diagnostics. If we want 32 bit tests I'll spin up a VM to generate those tests.

**Comment by alexcrichton** [comment]
ping r? @nikomatsakis 


Or @rust-lang/compiler I think Niko's out of town, so if someone else can take a look that'd be great!

**Comment by nikomatsakis** [comment]
Actually, I have some mild reservations. I just want to be sure: are those tests checking anything important? e.g., do we have other tests that show that the size of points on 32-bit machines is correctly handled, etc?

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum 

> I personally don't know if it's worth testing on those, the tests are just for diagnostics.

I guess my question is -- are there other tests that test similar conditions and *do* run on 32-bit?

**Comment by Mark-Simulacrum** [comment]
Yes, there are tests (the compile-fail tests that this edits, but doesn't move to UI, for one) that test all cases as far as I can tell. If you'd like, we can wait on this (probably closing it for now) until we get some form of ERROR-x86 or something like that so we can properly test across both platforms easily.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum ok, cool :)

**Comment by Mark-Simulacrum** [comment]
@bors rollup-

Since it failed CI last time.

**Pull request merged**